build-tools/packages/build-cli/src/commands/generate/releaseNotes.ts

```diff
@@ -0,0 +1,202 @@
+	UNKNOWN_SECTION,
+	difference,
+	groupBySection,
+	loadChangesets,
+} from "../../library/index.js";
+
+/**
+ * Generates release notes from individual changeset files.
+ */
+export default class GenerateReleaseNotesCommand extends BaseCommand<
+	typeof GenerateReleaseNotesCommand
+> {
+	static readonly summary = `Generates release notes from individual changeset files.`;
+
+	// Enables the global JSON flag in oclif.
+	static readonly enableJsonFlag = true;
+
+	static readonly flags = {
+		releaseGroup: releaseGroupFlag({
+			required: true,
+		}),
+		releaseType: Flags.custom<"major" | "minor">({
+			char: "t",
+			description: "The type of release for which the release notes are being generated.",
+			options: ["major", "minor"],
+			required: true,
+			parse: async (input) => {
+				if (input === "major" || input === "minor") {
+					return input;
+				}
+
+				throw new Error(`Invalid release type: ${input}`);
+			},
+		})(),
+		out: Flags.file({
+			description: `Output the results to this file.`,
+			required: true,
+			default: "RELEASE_NOTES.md",
+			// exists: true
+		}),
+		...BaseCommand.flags,
+	} as const;
+
+	static readonly examples = [
+		{
+			description: `Generate release notes for a minor release of the client release group.`,
+			command: "<%= config.bin %> <%= command.id %> -g client -t minor",
+		},
+		{
+			description: `You can output a different file using the --out flag.`,
+			command:
+				"<%= config.bin %> <%= command.id %> -g client -t minor --out RELEASE_NOTES/2.1.0.md",
+		},
+	];
+
+	public async run(): Promise<string> {
+		const context = await this.getContext();
+		const { flags, logger } = this;
+
+		const releaseGroup = context.repo.releaseGroups.get(flags.releaseGroup);
+		if (releaseGroup === undefined) {
+			this.errorLog(`Unknown release group: ${flags.releaseGroup}`);
+			this.exit(2);
+		}
+
+		const { releaseNotes: releaseNotesConfig } = loadFluidBuildConfig(
+			context.gitRepo.resolvedRoot,
+		);
+		if (releaseNotesConfig === undefined) {
+			this.errorLog(
+				`No releaseNotes config found. Make sure the 'releaseNotes' section of the build config exists.`,
+			);
+			this.exit(2);
+		}
+
+		const changesetDir = path.join(releaseGroup.directory, DEFAULT_CHANGESET_PATH);
+		const changesets = await loadChangesets(changesetDir, logger);
+
+		const { version } = releaseGroup;
+		const header = `<!-- THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. -->`;
+		const footer = `### 🛠️ Start Building Today!\n\nPlease continue to engage with us on GitHub
+[Discussion](https://github.com/microsoft/FluidFramework/discussions) and
+[Issue](https://github.com/microsoft/FluidFramework/issues) pages as you adopt Fluid Framework!
+`;
+		const intro = `# Fluid Framework v${version}\n\n## Contents`;
+
+		this.info(`Loaded ${changesets.length} changes.`);
+		assert(flags.releaseType !== undefined, `Release type must be provided.`);
+
+		const bySection = groupBySection(changesets);
+
+		{
```

Comment:
Is the "floating" code block deliberate?

Reply:
Might be worth a quick code comment if so :)

Reply:
This is something I picked up from @CraigMacomber and @jason-ha, but I am not sure this is a good place to use it. This
creates a new scope, and the variables defined within it drop out of scope at the end of the block. Since this block is sort
of self-contained, I thought it made sense. I suppose an alternative is to move this to a function, but then you have
to pass stuff around - not sure that's better.

Reply:
I had never thought of using them like that, basically as an alternative to functions that avoids arg passing or closures (maybe this technically still closes over the things outside the block, but it being inline with the rest of the code makes it feel different than a closure, IMO). A function defined here (letting it use the stuff outside it, not passing everything as args) might give the same dev experience, and in that case I would definitely think "so what is this closing over and how can that be a bug"; not sure if that or explicit block for scope is preferable. I'm fine keeping it, but would add a comment as Josh suggested, just for clarity.

Reply:
Floating blocks can be nice, but for another option to consider you can always make the block a named function and call it, or make it an IIFE, since sometimes giving the block a name (and a frame on the callstack in case of an error) is nice.